provider/aws/aws.go

```diff
@@ -324,69 +323,66 @@ func (p *AWSProvider) zones(ctx context.Context) (map[string]*profiledZone, erro
 	log.Debug("Refreshing zones list cache")
 
 	zones := make(map[string]*profiledZone)
-	var profile string
-	var tagErr error
-	f := func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool) {
-		for _, zone := range resp.HostedZones {
-			if !p.zoneIDFilter.Match(aws.StringValue(zone.Id)) {
-				continue
-			}
 
-			if !p.zoneTypeFilter.Match(zone) {
-				continue
-			}
+	for profile, client := range p.clients {
+		paginator := route53.NewListHostedZonesPaginator(client, &route53.ListHostedZonesInput{})
+		for paginator.HasMorePages() {
+			var tagErr error
 
-			if !p.domainFilter.Match(aws.StringValue(zone.Name)) {
-				if !p.zoneMatchParent {
+			resp, err := paginator.NextPage(ctx)
+			if err != nil {
+				var te *route53types.ThrottlingException
+				if errors.As(err, &te) {
+					log.Warnf("Skipping AWS profile %q due to provider side throttling: %v", profile, te.ErrorMessage())
 					continue
 				}
-				if !p.domainFilter.MatchParent(aws.StringValue(zone.Name)) {
+				// nothing to do here. Falling through to general error handling
+				return nil, provider.NewSoftError(fmt.Errorf("failed to list hosted zones: %w", err))
+			}
+			for _, zone := range resp.HostedZones {
+				if !p.zoneIDFilter.Match(*zone.Id) {
 					continue
 				}
-			}
 
-			// Only fetch tags if a tag filter was specified
-			if !p.zoneTagFilter.IsEmpty() {
-				tags, err := p.tagsForZone(ctx, *zone.Id, profile)
-				if err != nil {
-					tagErr = err
-					return false
-				}
-				if !p.zoneTagFilter.Match(tags) {
+				if !p.zoneTypeFilter.Match(zone) {
 					continue
 				}
-			}
 
-			zones[aws.StringValue(zone.Id)] = &profiledZone{
-				profile: profile,
-				zone:    zone,
-			}
-		}
+				if !p.domainFilter.Match(*zone.Name) {
+					if !p.zoneMatchParent {
+						continue
+					}
+					if !p.domainFilter.MatchParent(*zone.Name) {
+						continue
+					}
+				}
 
-		return true
-	}
+				// Only fetch tags if a tag filter was specified
+				if !p.zoneTagFilter.IsEmpty() {
+					tags, err := p.tagsForZone(ctx, *zone.Id, profile)
+					if err != nil {
+						tagErr = err
+						break
```

Comment:
This break does not breakout the right loop as far as I understand. If you check https://github.com/kubernetes-sigs/external-dns/blob/master/provider/aws/aws.go#L348C4-L353C18 and https://github.com/kubernetes-sigs/external-dns/blob/master/provider/aws/aws.go#L383
You likely need to continue to the outermost for loop that loops over p.client

Reply:
Oh good catch, I agree this is a deviation. While thinking about it more, do we want to keep the existing behavior? Both the original and current PR implementation start by looping over all profiles, all their hosted zone pages, and all hosted zones per page, populating a `zones` cache and keeping track of `tagErr`. Then,

Original: After looping through all hosted zones, if `tagErr != nil`, then return an error (and throw away the otherwise complete`zones` cache)

Current PR implementation: When the first `tagErr` is found, then return an error (and throw away however much progress was made in the `zones` cache)

I want to emphasize that this was an unintentional change, but if you agree, I may have improved it out of coincidence? Happy to make the logic match the original if you think that is best.

Reply:
I think for a refactoring we should aim for "no logic changes".
If you think it is an improvement then let's create another PR to improve the code, after the refactoring PR was merged.

Reply:
Sure thing, shifted the implementation to match the logic before the refactor - thanks!